crates/client/l1-messages/src/worker.rs

```diff
@@ -18,51 +15,50 @@ use crate::contract::{L1Contract, LogMessageToL2Filter};
 use crate::error::L1MessagesWorkerError;
 
 const TX_SOURCE: TransactionSource = TransactionSource::External;
-const STORAGE_PREFIX: &str = "L1MessagesWorker";
-const STORAGE_KEY: &str = "last_synced_block";
 
 pub fn connect_to_l1_contract(
     config: &L1MessagesWorkerConfig,
 ) -> Result<L1Contract<Provider<Http>>, L1MessagesWorkerError> {
-    let address: Address =
-        config.contract_address.parse::<Address>().map_err(|_| L1MessagesWorkerError::ConfigError)?;
-    let provider = Provider::<Http>::try_from(&config.http_provider).map_err(|e| {
+    let provider = Provider::<Http>::try_from(config.get_provider()).map_err(|e| {
         log::error!("⟠ Failed to connect to L1 Node: {:?}", e);
         L1MessagesWorkerError::ConfigError
     })?;
 
-    let l1_contract = L1Contract::new(address, Arc::new(provider));
+    let l1_contract = L1Contract::new(*config.get_contract_address(), Arc::new(provider));
 
     Ok(l1_contract)
 }
 
-pub async fn run_worker<C, P, B, S>(config: L1MessagesWorkerConfig, client: Arc<C>, pool: Arc<P>, backend: Arc<S>)
-where
+pub async fn run_worker<C, P, B>(
+    config: L1MessagesWorkerConfig,
+    client: Arc<C>,
+    pool: Arc<P>,
+    backend: Arc<mc_db::Backend<B>>,
+) where
     B: BlockT,
     C: ProvideRuntimeApi<B> + HeaderBackend<B>,
     C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
     P: TransactionPool<Block = B> + 'static,
-    S: Backend<B> + Send + 'static,
 {
     log::info!("⟠ Starting L1 Messages Worker with settings: {:?}", config);
 
-    let mut offchain_storage = match backend.offchain_storage() {
-        Some(offchain_storage) => offchain_storage,
-        None => {
-            log::error!("⟠ Offchain storage unavailable");
+    let l1_contract = match connect_to_l1_contract(&config) {
+        Ok(l1_contract) => l1_contract,
+        Err(e) => {
+            log::error!("⟠ Unexpected error while connecting to L1: {:?}", e);
             return;
         }
     };
 
-    let l1_contract = match connect_to_l1_contract(&config) {
-        Ok(l1_contract) => l1_contract,
+    let last_synced_l1_block = match backend.messaging().last_synced_l1_block() {
+        Ok(blknum) => blknum,
         Err(e) => {
-            log::error!("⟠ Unexpected error while connecting to L1: {:?}", e);
+            log::error!("⟠ Madara Messaging DB unavailable: {:?}", e);
             return;
         }
     };
 
-    let events = l1_contract.events().from_block(get_block_number::<B, S>(&offchain_storage).unwrap_or_default());
+    let events = l1_contract.events().from_block(last_synced_l1_block);
```

Comment:
```
let events_stream = l1_contract.events().from_block(last_synced_l1_block).steam();
```

It will be more expressive than this one that I struggled to understand.

Reply:
you can even use https://docs.rs/ethers/latest/ethers/middleware/contract/struct.Event.html#method.stream_with_meta directly

Reply:
Unfortunately, the compiler says `creates a temporary value which is freed while still in use` for `l1_contract.events().from_block(last_synced_l1_block).stream().await` . We need to have `events` assigned just like in this example - https://docs.rs/ethers/latest/ethers/middleware/contract/struct.Event.html#method.stream